api/src/pcapi/core/users/api.py

```diff
@@ -471,6 +475,112 @@ def bulk_unsuspend_account(user_ids: list[int], actor: models.User) -> None:
     )
 
 
+class RollbackableCommand:
+    """A command that can be rolled back"""
+
+    def __init__(
+        self,
+        command: typing.Callable,
+        rollback: typing.Callable,
+        set_initial_state: typing.Callable = None,
+    ) -> None:
+        self.command_func = command
+        self.rollback_func = rollback
+        self.set_initial_state_func = set_initial_state
+        self.initial_state = None
+
+    def execute(self) -> None:
+        """Execute the command and save the initial state if needed"""
+        self.initial_state = self.set_initial_state_func() if self.set_initial_state_func else None
+        self.command_func()
+
+    def rollback(self) -> None:
+        """Rollback the command to its initial state"""
+        if self.initial_state:
+            self.rollback_func(self.initial_state)
+
+
+class Transaction:
+    """A transaction is a list of commands that can be executed or rolled back
+    in case of failure"""
+
+    def __init__(self) -> None:
+        self.commands: typing.List[RollbackableCommand] = []
+        self.current_command_index = 0
+
+    def add_command(self, command: RollbackableCommand) -> None:
+        """Add a command to the transaction"""
+        self.commands.append(command)
+
+    def execute(self) -> None:
+        """Execute all the commands in the transaction
+        If one command fails, the transaction is rolled back"""
+        try:
+            for command in self.commands:
+                command.execute()
+                self.current_command_index += 1
+        except Exception as exception:
+            for i in range(self.current_command_index - 1, -1, -1):
+                try:
+                    self.commands[i].rollback()
+                except Exception as rollback_exception:  # pylint: disable=broad-except
+                    logger.error(
+                        "Rollback failed",
+                        extra={
+                            "command": self.commands[i],
+                            "exception": rollback_exception,
+                        },
+                    )
+            raise exception
```

Comment:
[pinaillage]
juste une suggestion pour cette méthode qui se passe de `self.current_command_index` tout en étant un poil plus lisible 
```suggestion
    def execute(self) -> None:
        """Execute all the commands in the transaction
        If one command fails, the transaction is rolled back"""
        executed_commands = []
        try:
            for command in self.commands:
                command.execute()
                executed_commands.append(command)
        except Exception as exception:
            for command in executed_commands:
                try:
                    command.rollback()
                except Exception as rollback_exception:  # pylint: disable=broad-except
                    logger.error(
                        "Rollback failed",
                        extra={
                            "command": self.commands[i],
                            "exception": rollback_exception,
                        },
                    )
            raise exception
```